commit-graph.c

```diff
@@ -109,8 +109,14 @@ timestamp_t commit_graph_generation(const struct commit *c)
 		return GENERATION_NUMBER_V2_INFINITY;
 	else if (data->graph_pos == COMMIT_NOT_FROM_GRAPH)
 		return GENERATION_NUMBER_V2_INFINITY;
+	else if (data->odate)
+		/*
+		 * MSB differentiates between zero offset and missing
+		 * generation data chunk
+		 */
+		return c->date + (data->odate & 0x7FFFFFFF);
 
-	return data->generation;
+	return data->level;
```

Comment:
I think `data` should just have a `uint64_t generation`, and you fill that value with the v2 generation number (`date + offset`) when parsing the commit OR fill it with the v1 generation number if `GDAT` doesn't exist (again, when parsing the commit). That makes all of your other edits unnecessary, as we can just compare the `generation` directly and know we are correct one way or another.